typing/typecore.ml

```diff
@@ -5453,9 +5453,20 @@ and type_expect_
             alloc_mode;
           }
         | Unboxed_product ->
+          let opt_exp = match opt_exp with
+            | None -> None
+            | Some (exp, _) ->
+              let jkind =
+                Jkind.Builtin.product ~why:Unboxed_record
+                  (Array.map (fun lbl -> lbl.lbl_jkind) label_descriptions
+                   |> Array.to_list)
+              in
+              let sort = Jkind.sort_of_jkind jkind in
+              Some (exp, sort)
+          in
```

Comment:
As in the previous commit, I think we should get the sort from `exp` rather than building a new one from the declaration here.  Probably the code that computes the typed version of `opt_exp` nearer the top of this function should be doing that - that's the right place to check the value we're overriding is representable (which we'll need to do anyway when we go from checking records are representable at declaration time to checking they are representable at construction time).

Reply:
Updated in 7ed17a022d34ae4d60f1a7205aba0f7c329b950!

Reply:
Sorry, I think this is still not quite right.  You are computing the sort from `ty_expected`, which I believe is the type of the result record.  But in a world where we allow any in declarations, these may not be the same. E.g.:
```ocaml
type 'a t = { x : 'a; y : string }
let update_sort (r : int64# t) : float# t = { r with x = Int64_u.of_float r.x }
```
You can use `exp.exp_type`, or whatever other copies of the type of the original record are lying around here.

Reply:
Oops, thanks. Updated in c0382c6d35e4ec4593069c4b2da8a1ad15634dcf